### PR TITLE
obs-transitions: Add color to 'Luma Wipe' Transitions

### PR DESCRIPTION
**Pull request by Oncorporation**
Add color to Luma Wipes with softness
Fix divide by zero softness
Use black for no color
Add OBS Logo Wipes and 4 other derezzed style Wipes - personally created

### CODE REVIEW

**Comment by Vainock** [comment]
Why have you defined the paths in the locale file? Does not make sense for me.

**Comment by Gol-D-Ace** [comment]
Hi

The strings should contain the effect name and not the image name like Vainock already mentioned.

**Comment by Oncorporation** [comment]
I just updated the code that has been around for a while. I always follow the current architecture of the project.

**Comment by Vainock** [comment]
https://cdn.discordapp.com/attachments/549677588143013898/555420043501109259/Screenshot_20190313-170038.png

**Comment by Oncorporation** [comment]
> https://cdn.discordapp.com/attachments/549677588143013898/555420043501109259/Screenshot_20190313-170038.png

Answering your own question?  I like😁

**Comment by Vainock** [comment]
? What do you mean? I only provided a picture where you can see that you made something wrong. It's not correct that the file names are in the drop down list...

**Comment by Oncorporation** [comment]
Oh. I thought you were saying that the actual placement of where the strings were was wrong.  Will update the config file, thank you.

**Comment by Vainock** [comment]
The placement is wrong. The locale files are only for the texts that should be shown in the GUI, not for file names or something like that.

**Comment by Oncorporation** [comment]
These string are shown in the GUI. The actual file names are in Wipes.json. 

I had copied and pasted to ensure everything lined up while developing and did not go back and change the names for the GUI drop down list. So, thank you for your review!

This is now updated, but I did not rebase and squash for a settings file.

File: plugins/obs-transitions/transition-luma-wipe.c [hunk 1]
```diff
@@ -35,11 +39,16 @@ static const char *luma_wipe_get_name(void *type_data)
 
 static void luma_wipe_update(void *data, obs_data_t *settings)
 {
-	struct luma_wipe_info *lwipe = data;
+	struct luma_wipe_info *lwipe = data;
```
Comment (RytoEX) [inline plugins/obs-transitions/transition-luma-wipe.c]: Style:  Don't add trailing whitespace.

```suggestion
	struct luma_wipe_info *lwipe = data;
```

File: plugins/obs-transitions/transition-luma-wipe.c [hunk 2]
```diff
@@ -98,9 +107,11 @@ static void *luma_wipe_create(obs_data_t *settings, obs_source_t *source)
 	lwipe->ep_a_tex    = gs_effect_get_param_by_name(effect, "a_tex");
 	lwipe->ep_b_tex    = gs_effect_get_param_by_name(effect, "b_tex");
 	lwipe->ep_l_tex    = gs_effect_get_param_by_name(effect, "l_tex");
+	lwipe->ep_l_color = gs_effect_get_param_by_name(effect, "l_color");
 	lwipe->ep_progress = gs_effect_get_param_by_name(effect, "progress");
 	lwipe->ep_invert   = gs_effect_get_param_by_name(effect, "invert");
 	lwipe->ep_softness = gs_effect_get_param_by_name(effect, "softness");
+
```
Comment (RytoEX) [inline plugins/obs-transitions/transition-luma-wipe.c]: Don't make unnecessary whitespace changes.
```suggestion
```

File: plugins/obs-transitions/transition-luma-wipe.c [hunk 3]
```diff
@@ -98,9 +107,11 @@ static void *luma_wipe_create(obs_data_t *settings, obs_source_t *source)
 	lwipe->ep_a_tex    = gs_effect_get_param_by_name(effect, "a_tex");
 	lwipe->ep_b_tex    = gs_effect_get_param_by_name(effect, "b_tex");
 	lwipe->ep_l_tex    = gs_effect_get_param_by_name(effect, "l_tex");
+	lwipe->ep_l_color = gs_effect_get_param_by_name(effect, "l_color");
```
Comment (RytoEX) [inline plugins/obs-transitions/transition-luma-wipe.c]: Style:  Align the assignment operator (`=`) with the lines above and below it.

```suggestion
	lwipe->ep_l_color  = gs_effect_get_param_by_name(effect, "l_color");
```

File: plugins/obs-transitions/transition-luma-wipe.c [hunk 4]
```diff
@@ -5,10 +5,12 @@
 #define S_LUMA_IMG              "luma_image"
 #define S_LUMA_INV              "luma_invert"
 #define S_LUMA_SOFT             "luma_softness"
+#define S_LUMA_COLOR		"luma_color"
```
Comment (RytoEX) [inline plugins/obs-transitions/transition-luma-wipe.c]: Style:  Use spaces for alignment instead of tabs.

```suggestion
#define S_LUMA_COLOR            "luma_color"
```

File: plugins/obs-transitions/data/luma_wipe_transition.effect [hunk 5]
```diff
@@ -35,22 +37,31 @@ float4 PSLumaWipe(VertData v_in) : TARGET
...
 
-	float time = lerp(0.0f, 1.0f + softness, progress);
+	// user color with luma
+	float4 out_color = float4(l_color.rgb, luma);
+
+	float time = lerp(0.0f, 1.0f + soft, progress);
 
-	if (luma <= time - softness)
-		return b_color;
+	if (luma <= time - soft){
```
Comment (RytoEX) [inline plugins/obs-transitions/data/luma_wipe_transition.effect]: Style:  Place a space between the closing parenthesis and the opening curly brace.

```suggestion
	if (luma <= time - soft) {
```

**Comment by DDRBoxman** [comment]
This might be better split into two PRs one with the color changes. And then one that isn't adding more wipes but a way to load custom wipes. Then users can post luma wipes on the resources section of the forum.

Also I'd like to see the color options on action with an example.

**Comment by Xaymar** [comment]
### Commit 51673d9b900509156ed31157a9add2ad70d9aa3a
This commit seems to be doing more than it describes, it also adds new luma wipe filters. This should be split into multiple commits instead of all being one commit, one for fixing the divide issue, one for color support and one for the addition of new luma wipe filters.

### Commit dbb0b808293e644014d9ca5c045bd51ffc4d7022
Should be squashed into the above addition of new luma wipe filters.

### Commit 797193a1bf967d49f72d7d40c11db5aae1f861fa
Should also be squished into the above luma wipe filters. The title of the commit is also non-compliant with the general style of the rest of the commits in obs-studio.

File: plugins/obs-transitions/data/luma_wipe_transition.effect [hunk 6]
```diff
@@ -35,22 +37,31 @@ float4 PSLumaWipe(VertData v_in) : TARGET
 	float2 uv = v_in.uv;
 	float4 a_color = convert_pmalpha(a_tex.Sample(textureSampler, uv));
 	float4 b_color = convert_pmalpha(b_tex.Sample(textureSampler, uv));
-	float luma     = l_tex.Sample(textureSampler, uv).x;
+	float luma = l_tex.Sample(textureSampler, uv).x;
+	float soft = softness;
+
+	if (soft == 0.00)
+		soft += 0.01;
```
Comment (Xaymar) [inline plugins/obs-transitions/data/luma_wipe_transition.effect]: Consider using clamp or min/max instead of if. Newer drivers _might_ optimize it into the same instructions, but older drivers will suffer from ifs that rely on uniforms.

**Comment by Oncorporation** [comment]
> This might be better split into two PRs one with the color changes. And then one that isn't adding more wipes but a way to load custom wipes. Then users can post luma wipes on the resources section of the forum.
> 
> Also I'd like to see the color options on action with an example.
I made this at like 4 am ... hehe but I did go through it to show people how to use it
https://www.twitch.tv/videos/394810713?t=00h04m55s

File: plugins/obs-transitions/data/luma_wipe_transition.effect (hunk 6 above)
Comment (Oncorporation) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: clamp would be better

Comment (Xaymar) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: The problem is that if you use uniforms anywhere in the calculation, even anywhere earlier, and don't do some limiting on the uniform variable so the shader compiler has a known bound, you will get the same compilation route as just directly using the uniform. So these two have different compilation paths:
```
uniform float softness;
float soft = softness;
if (soft == 0.00) {
    soft += 0.001;
}
```
```
uniform float softness;
float soft = max(softness, 1.0e-6);
```
The first one will simply be an if, or perhaps converted to a more optimal check against 0, while the second one actually places a lower bound on soft (but no upper bound). Having any restrictive bounds can lead to better optimized final shader code than having no bounds. Though in this case it might not really make a difference on modern cards, but older cards, especially those struggling with ifs and jumps will see a difference.

Can softness be negative, or just a positive value of 0 to infinity?
```
uniform float softness;
float soft = max(softness, 1.0e-6);
```
Or is it -infinity to +infinity? In this case, abs and sign can be used.
```
uniform float softness;
float soft = max(abs(softness), 1.0e-6) * sign(softness);
```

Ideally this limiting would not be done on the GPU, as it can be done on the CPU at much lower cost.

Comment (Oncorporation) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: Thanks for the explanation. I was under the impression that converting variables away from uniform was enough to mitigate some of these issues.

I do like the max * sign, am going to test and probably implement.

**Comment by Oncorporation** [comment]
Squashed

File: plugins/obs-transitions/data/luma_wipe_transition.effect (hunk 6 above)
Comment (Xaymar) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: Consider simply doing this on the CPU, or at least caching the result in the vertex shader. Doing this calculation per fragment is going to relatively kill performance.

Comment (Oncorporation) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: Are you referring to the soft definition or the entire operation. On average the operation only affect a narrow range of pixels equal to about twice the softness range.

Comment (Xaymar) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: I'm referring to the limiting of softness to not be 0, which is not a necessary per-fragment or per-vertex operation. The fragment shader is executed for every fragment, which means that your limiting happens 124416000 times per second at 1920x1080@60fps, compared to just 60 times if you did it on the CPU. That is rather huge difference that actually matters in situations where the GPU is used heavily by game.

Comment (Oncorporation) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: I agree, the placement of that calculation does not need to be per fragment. Making an adjustment

**Comment by Oncorporation** [comment]
Squashed Again

File: plugins/obs-transitions/data/luma_wipe_transition.effect (hunk 6 above)
Comment (Oncorporation) [reply plugins/obs-transitions/data/luma_wipe_transition.effect]: Moved into the interface, not allowing 0 and only using positive numbers fir softness

**Comment by jp9000** [comment]
Hey there, thank you again for your proposed changes.  For this PR, I'm not sure how I feel about it.  The logo wipe is a nice little tribute to the program, although the problem with it is that it kind of gets stretched, so it doesn't really look that good most of the time due to the fact that the program usually runs in 16:9 resolution.

For the color change, I personally just am not feeling it very much.  It feels kind of a little bit lackluster.  It feels a bit weird and unusual to me.  Perhaps that's just me, but personally I just feel like this PR is not necessary, so I must respectfully decline this PR.  I hope you don't mind that fact, and I know that it can be frustrating for a maintainer to disagree on feature changes/additions.  Please don't let that stop you from contributing.

**Comment by Oncorporation** [comment]
Again, I wish you would discuss these things with me. There are reasons for the decisions I make that need explained more. Maybe I will resubmit with more explanation.
